### PR TITLE
Fix for flaky test.

### PR DESCRIPTION
**Pull request by pstibrany**
Test failure reported by @gouthamve, after recent merge of PR #1756:

```
--- FAIL: TestWatchPrefix (3.86s)
    --- FAIL: TestWatchPrefix/etcd (1.86s)
        kv_test.go:191: key test/97 has incorrect value 0
        kv_test.go:191: key test/98 has incorrect value 0
        kv_test.go:191: key test/99 has incorrect value 0
```

~Test was too strict. Now it checks that we receive most notifications, but not necessarily all.~

Test watches for changes under prefix "test/" and expect single change in every key between 0-100. 
It failed because it didn't wait long enough. Now it waits for longer time, but stops as soon it receives all notifications. Also watching for changes starts before generator starts, so that first changes can't possibly be missed.

### CODE REVIEW

File: pkg/ring/kv/kv_test.go [hunk 1]
```diff
@@ -183,18 +183,26 @@ func TestWatchPrefix(t *testing.T) {
...
 			}
+			count += observedKeys[key]
 			delete(observedKeys, key)
 		}
 
 		if len(observedKeys) > 0 {
 			t.Errorf("unexpected keys reported: %v", observedKeys)
 		}
+
+		const expectedFactor = 0.75
```
Comment (pracucci) [inline pkg/ring/kv/kv_test.go]: I'm not a big fan of this, cause it doesn't guarantee it works as expected (ie. no key has been missed). A trick may be waiting - right before the `for` until the number of observed keys doesn't increase for some time (ie. 500ms). You may do something like this:

```
		// wait until the WatchPrefix() stop gets called, which means all the
		// keys have been observed
		lastObservedKeysCount := len(observedKeys)
		lastObservedKeysChangeAt := time.Now()

		test.Poll(t, 2*time.Second, true, func() interface{} {
			count := len(observedKeys)
			elapsed := time.Now().Sub(lastObservedKeysChangeAt)

			if count == lastObservedKeysCount {
				return elapsed >= 250*time.Millisecond
			}

			lastObservedKeysCount = count
			lastObservedKeysChangeAt = time.Now()
			return false
		})
```

Does it make sense?

Comment (pstibrany) [reply pkg/ring/kv/kv_test.go]: I've tried to implement your suggestion. Test is strict again (each key once), but more forgiving in terms of timeouts. But if nothing is received in 500ms, we stop.

File: pkg/ring/kv/kv_test.go [hunk 2]
```diff
@@ -165,23 +178,35 @@ func TestWatchPrefix(t *testing.T) {
...
-		client.WatchPrefix(ctx, prefix, func(key string, val interface{}) bool {
-			observedKeys[key] = observedKeys[key] + 1
-			return true
-		})
 
-		// wait until updaters finish (should be done by now)
-		<-ch1
-		<-ch2
+		totalDeadline := time.After(totalTestTimeout)
+		var nextTimer = time.NewTimer(nextKeyDeadline)
```
Comment (pracucci) [inline pkg/ring/kv/kv_test.go]: To simplify it, what if we remove the `nextTimer` at all, and we simply set `watching = false` when `len(observedKeys)` reach `max`?

Comment (pstibrany) [reply pkg/ring/kv/kv_test.go]: Sounds good

**Comment by pstibrany** [comment]
> Thanks @pstibrany to patiently address my feedback. LGTM!

Thanks for your review and feedback!

**Pull request merged**